client/src/components/HistoryViewer/HistoryViewer.js

```diff
@@ -308,97 +257,121 @@ class HistoryViewer extends Component {
-    const {
-      isInGridField,
-      ListComponent,
-      CompareWarningComponent,
-      compare,
-      compare: { versionFrom: hasVersionFrom },
-    } = this.props;
-
+  const renderVersionList = () => {
+    const hasVersionFrom = compare.versionFrom || null;
     return (
-      <div className={this.getContainerClasses()}>
+      <div className={getContainerClasses()}>
         <CompareWarningComponent />
 
         <div className={isInGridField ? '' : 'panel panel--padded panel--scrollable'}>
-          {this.renderComparisonSelectionList()}
+          {renderComparisonSelectionList()}
           <ListComponent
-            versions={this.state.versions}
+            versions={versions}
             showHeader={!compare || (compare && !hasVersionFrom)}
-            compareModeAvailable={this.compareModeAvailable()}
+            compareModeAvailable={compareModeAvailable()}
           />
 
           <div className="history-viewer__pagination">
-            {this.renderPagination()}
+            {renderPagination()}
           </div>
         </div>
       </div>
     );
-  }
-
-  renderCompareMode() {
-    const { compare } = this.props;
+  };
 
+  const renderCompareMode = () => {
     if (compare && compare.versionFrom && compare.versionTo) {
-      return this.renderVersionDetail();
+      return renderVersionDetail();
     }
-    return this.renderVersionList();
-  }
-
-  render() {
-    const { graphQLErrors, loading, compare, currentVersion, recordId } = this.props;
+    return renderVersionList();
+  };
 
-    // Do not proceed if there are graphql error.
-    // A toast message will be shown in componentDidMount() or componentDidUpdate()
+  /**
+   * Display a toast if there were any pre-existing graphql errors
+   */
+  useEffect(() => {
+    refreshVersionData();
     if (graphQLErrors.length > 0) {
-      return null;
+      toastActions.error(i18n._t('Admin.UNKNOWN_ERROR', 'An unknown error has occurred.'));
     }
+    prevErrorCountRef.current = graphQLErrors.length;
+  }, []);
 
-    if (!recordId) {
-      return null;
+  /**
+   * Display a toast if there were any new graphql errors
+   */
+  useEffect(() => {
+    if (graphQLErrors.length > prevErrorCountRef.current) {
+      toastActions.error(i18n._t('Admin.UNKNOWN_ERROR', 'An unknown error has occurred.'));
     }
+    prevErrorCountRef.current = graphQLErrors.length;
+  }, [graphQLErrors]);
 
-    if (loading) {
-      return <Loading />;
+  /**
+   * Refresh version data when the page changes
+   */
+  useEffect(() => {
+    if (versions.length === 0) {
+      return;
     }
+    refreshVersionData();
+  }, [page]);
 
-    if (this.compareModeAvailable() && compare) {
-      return this.renderCompareMode();
+  /**
+   * Reset the selected version when unmounting HistoryViewer to prevent data leaking
+   * between instances
+   */
+  useEffect(() => () => {
+    if (typeof onSelect === 'function') {
+      onSelect(0);
     }
+  }, [onSelect]);
```

Comment:
This used to be in `componentWillUnmount` - shouldn't it be returned from the `useEffect` that has no dependencies?

Reply:
Updated